File: app/src/main/java/com/haruhi/bismark439/haruhiism/WallPaper/yukiWall.java
package com.haruhi.bismark439.haruhiism.WallPaper;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.os.Handler;
import android.service.wallpaper.WallpaperService;
import android.view.SurfaceHolder;

import com.haruhi.bismark439.haruhiism.R;

import java.util.Random;

import static com.haruhi.bismark439.haruhiism.WallPaper.yukiSetting.loadLWP;

/**
 * Created by Bismark439 on 02/03/2018.
 */

public class yukiWall extends WallpaperService{
    private boolean mVisible;  // visible flag
    Canvas canvas;      // canvas reference
    int Drawspeed=33;   // thread call delay time 10 default
    int changePicSpeed = 47000;
    Context mcontext;   //reference to the current context
    final Handler mHandler = new Handler(); // this is to handle the thread
 int red = Color.parseColor("#CD5C5C");
    @Override
    public Engine onCreateEngine() {
        mcontext = this;  //set the current context
        //return the Engine Class
        return new LiveWall(); // this calls contain the wallpaper code
    }

    class LiveWall extends Engine
    {
        private final Runnable mDrawFrame = new Runnable() {
            public void run() {
                drawFrame();
            }
        };
        private final Runnable mChangePic = new Runnable() {
            public void run() {
               changePicture();
            }
        };
        // ======== MATRIX LIVE WALLPAPER VARS
        int background_color= Color.parseColor("#FF000000");
        private int width = 1000000; //default initial width
        private int height = 100; //default initial height


        int columnSize = 100; //column size ; no of digit required to fill the screen
        int fontSize = 0;//24; //font size of the text which will fall

        Paint textPaint;
        Paint backPaint;
        String text ="SUZUMIYAHARUHI";// "涼宮ハルヒ";  // Text which need to be drawn
        char[] textChar = text.toCharArray(); // split the character of the text
        int textLength = textChar.length;   //length of the length text

        int[]  textPosition; // contain the position which will help to draw the text
        int text_color=Color.parseColor("#FF8BFF4A");
        Random rand = new Random(); //random generater
        int[]imgList = {R.string.haruhi_ascii3,R.string.ascii_mikuru,R.string.ascii_hrhLN,R.string.ascii_haruhi3,R.string.ascii_sos};
        int mImgIndex = 0;
        char[][] foregroundImage;
        CharSequence asciiString;
        int imageHeight;
        int ImageMultiply = 2;
        //======================

        //Called when the surface is created
        @Override
        public void onSurfaceCreated(SurfaceHolder holder) {
            super.onSurfaceCreated(holder);
        }

        // remove thread
        @Override
        public void onDestroy() {
            super.onDestroy();
            mHandler.removeCallbacks(mDrawFrame);
            mHandler.removeCallbacks(mChangePic);
        }


        //called when varaible changed
        @Override
        public void onVisibilityChanged(boolean visible) {
            mVisible = visible;
            System.out.println("VisChangeCalled"+visible);
            text = loadLWP(getApplicationContext());
            textChar = text.toCharArray(); // split the character of the text
            textLength = textChar.length;
            fontSize = width/columnSize;
            if (visible) {
                drawFrame();
                changePicture();
            } else {
                //this is necessay to remove the call back
                mHandler.removeCallbacks(mDrawFrame);
                mHandler.removeCallbacks(mChangePic);
            }
        }

        //called when surface destroyed
        @Override
        public void onSurfaceDestroyed(SurfaceHolder holder) {
            super.onSurfaceDestroyed(holder);
            mVisible = false;
            //this is necessay to remove the call back
            mHandler.removeCallbacks(mDrawFrame);
            mHandler.removeCallbacks(mChangePic);
        }


        //this function contain the the main draw call
        /// this function need to call every time the code is executed
        // the thread call this function with some delay "drawspeed"
        public void drawFrame()
        {
            //getting the surface holder
            final SurfaceHolder holder = getSurfaceHolder();
            canvas = null;  // canvas
            try {
                canvas = holder.lockCanvas();  //get the canvas
                if (canvas != null) {
                    // draw something
                    // canvas matrix draw code
                    canvasDraw();
                    //^^^^
                }
            } finally {
                if (canvas != null)
                    holder.unlockCanvasAndPost(canvas);
            }

            // Reschedule the next redraw
            // this is the replacement for the invialidate funtion
            // every time call the drawFrame to draw the matrix
            mHandler.removeCallbacks(mDrawFrame);
            if (mVisible) {
                // set the execution delay
                mHandler.postDelayed(mDrawFrame, Drawspeed);
            }
        }
        public void changePicture()
            {
                constructImage();
                mImgIndex = rand.nextInt(imgList.length);
                System.out.println("Rand pic "+mImgIndex);
//                if(mImgIndex>=imgList.length)mImgIndex=0;
                mHandler.postDelayed(mChangePic,changePicSpeed);
        }

        public void initialise() {
            textPaint = new Paint();
            backPaint = new Paint();
            textPaint.setStyle(Paint.Style.FILL);
            textPaint.setColor(text_color);
            textPaint.setTextSize(fontSize);

            backPaint.setColor(background_color);
            backPaint.setAlpha(5); //set the alpha
            backPaint.setStyle(Paint.Style.FILL);
        }
        @Override
        public void onSurfaceChanged(SurfaceHolder holder, int format, int width, int height) {
            super.onSurfaceChanged(holder, format, width, height);
            // some matrix variable
            // though not needed
            Paint paint = new Paint();
            paint.setColor(background_color);
            paint.setAlpha(255); //set the alpha
            paint.setStyle(Paint.Style.FILL);
            this.width = width + 20;
            this.height = height;
            System.out.println("onSurfaceChanged WIDTH: "+width+" HEIGHT "+height);
            fontSize = width/columnSize;
            constructImage();
            initialise();
            //initalise the textposiotn to zero
            textPosition = new int[columnSize+1]; //add one more drop
            for(int x = 0; x < columnSize; x++) {
                textPosition[x] = 1;
            }
            drawFrame();
            changePicture();
            canvas.drawRect(0, 0, width,height, paint);

        }
        //old matrix effect code
        void drawText()
        {
            //loop and paint
            for(int i =0 ;i<textPosition.length;i++)
            {
                // draw the text at the random position
                int x = i;
                int y = textPosition[i];
                if(x>=99 || y>= imageHeight * ImageMultiply ||foregroundImage[y][x]=='ㅡ'){
                    textPaint.setColor(Color.parseColor("#00c800"));
                    canvas.drawText("" + textChar[rand.nextInt(textLength)], x * fontSize, y * fontSize, textPaint);
                //
                }else{
                    textPaint.setColor(red);
                    //canvas.drawText(".", x * fontSize, y * fontSize, paint);
                    canvas.drawText(""+foregroundImage[y][x], x * fontSize, y * fontSize, textPaint);
                }

                // check if text has reached bottom or not
                //0.925 good
                if(textPosition[i]*fontSize > height && Math.random() > 0.900) //0.975
                    textPosition[i] = 0;   // change text position to zero when 0 when text is at the bottom

                textPosition[i]++; //increment the position array
            }

        }
        public void constructImage(){
            asciiString = getString(imgList[mImgIndex]);
            imageHeight = (asciiString.length() / 100);
            foregroundImage = new char [imageHeight * ImageMultiply][100];
          //  System.out.println("LENGTH "+asciiString.length());
            int modY = 0;
            for(int y = 0 ; y <imageHeight ; y++){
                for(int x = 0 ; x <100 ; x++){
                    foregroundImage[modY][x] = asciiString.charAt(x + y*100);
                    foregroundImage[modY+1][x] =  foregroundImage[modY][x];
                //    foregroundImage[modY+2][x] =  foregroundImage[modY][x];
                }
                modY = modY+ImageMultiply;
            }


        }
        //old martix effect code
        public void canvasDraw()
        {
            //Log.d("canvas ","drawing");
            //set the paint for the canvas
            canvas.drawRect(0, 0, width, height, backPaint);//draw rect to clear the canvas
            drawText(); // draw the canvas

        }

    }



}

File: app/src/main/java/com/haruhi/bismark439/haruhiism/wallPapers/YukiWall.kt
package com.haruhi.bismark439.haruhiism.wallPapers

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.os.Handler
import android.service.wallpaper.WallpaperService
import android.view.SurfaceHolder
import com.haruhi.bismark439.haruhiism.DEBUG
import com.haruhi.bismark439.haruhiism.R
import com.haruhi.bismark439.haruhiism.wallPapers.ActivityYukiLWPSettings.Companion.loadLWP
import java.util.*

/**
 * Created by Bismark439 on 02/03/2018.
 */

class YukiWallpaper : WallpaperService() {
    private var mVisible // visible flag
            = false
    var canvas // canvas reference
            : Canvas? = null
    var drawSpeed = 33 // thread call delay time 10 default
    var changePicSpeed = 47000
    private var mcontext //reference to the current context
            : Context? = null
    val mHandler = Handler() // this is to handle the thread
    var red = Color.parseColor("#CD5C5C")
    override fun onCreateEngine(): Engine {
        mcontext = this //set the current context
        //return the Engine Class
        return LiveWall() // this calls contain the wallpaper code
    }

    internal inner class LiveWall : Engine() {
        private val mDrawFrame = Runnable { drawFrame() }
        private val mChangePic = Runnable { changePicture() }

        // ======== MATRIX LIVE WALLPAPER VARS
        private var backgroundColor = Color.parseColor("#FF000000")
        private var width = 1000000 //default initial width
        private var height = 100 //default initial height
        private var columnSize = 100 //column size ; no of digit required to fill the screen
        private var fontSize = 0 //24; //font size of the text which will fall
        private var textPaint: Paint? = null
        private var backPaint: Paint? = null
        var text: String = "SUZUMIYAHARUHI" // "涼宮ハルヒ";  // Text which need to be drawn
        private var textChar = text.toCharArray() // split the character of the text
        private var textLength = textChar.size //length of the length text

        // contain the position which will help to draw the text
        private lateinit var textPosition: IntArray
        private var textColor = Color.parseColor("#FF8BFF4A")
        private var rand = Random() //random generater
        private var imgList = intArrayOf(
            R.string.haruhi_ascii3,
            R.string.ascii_mikuru,
            R.string.ascii_hrhLN,
            R.string.ascii_haruhi3,
            R.string.ascii_sos
        )
        private var mImgIndex = 0
        private lateinit var foregroundImage: Array<CharArray>
        private var asciiString: CharSequence? = null
        private var imageHeight = 0
        private var imageMultiply: Int = 2

        //======================
        //Called when the surface is created
        override fun onSurfaceCreated(holder: SurfaceHolder) {
            super.onSurfaceCreated(holder)
        }

        // remove thread
        override fun onDestroy() {
            super.onDestroy()
            mHandler.removeCallbacks(mDrawFrame)
            mHandler.removeCallbacks(mChangePic)
        }

        //called when varaible changed
        override fun onVisibilityChanged(visible: Boolean) {
            mVisible = visible
            text = loadLWP(applicationContext)
            textChar = text.toCharArray() // split the character of the text
            textLength = textChar.size
            fontSize = width / columnSize
            if (visible) {
                drawFrame()
                changePicture()
            } else {
                //this is necessay to remove the call back
                mHandler.removeCallbacks(mDrawFrame)
                mHandler.removeCallbacks(mChangePic)
            }
        }

        //called when surface destroyed
        override fun onSurfaceDestroyed(holder: SurfaceHolder) {
            super.onSurfaceDestroyed(holder)
            mVisible = false
            //this is necessay to remove the call back
            mHandler.removeCallbacks(mDrawFrame)
            mHandler.removeCallbacks(mChangePic)
        }

        //this function contain the the main draw call
        /// this function need to call every time the code is executed
        // the thread call this function with some delay "drawspeed"
        private fun drawFrame() {
            //getting the surface holder
            val holder = surfaceHolder
            canvas = null // canvas
            try {
                canvas = holder.lockCanvas() //get the canvas
                if (canvas != null) {
                    // draw something
                    // canvas matrix draw code
                    canvasDraw()
                    //^^^^
                }
            } finally {
                if (canvas != null) holder.unlockCanvasAndPost(canvas)
            }

            // Reschedule the next redraw
            // this is the replacement for the invialidate funtion
            // every time call the drawFrame to draw the matrix
            mHandler.removeCallbacks(mDrawFrame)
            if (mVisible) {
                // set the execution delay
                mHandler.postDelayed(mDrawFrame, drawSpeed.toLong())
            }
        }

        private fun changePicture() {
            constructImage()
            mImgIndex = rand.nextInt(imgList.size)
            mHandler.postDelayed(mChangePic, changePicSpeed.toLong())
        }

        private fun initialise() {
            textPaint = Paint()
            backPaint = Paint()
            textPaint!!.style = Paint.Style.FILL
            textPaint!!.color = textColor
            textPaint!!.textSize = fontSize.toFloat()
            backPaint!!.color = backgroundColor
            backPaint!!.alpha = 5 //set the alpha
            backPaint!!.style = Paint.Style.FILL
        }

        override fun onSurfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
            super.onSurfaceChanged(holder, format, width, height)
            // some matrix variable
            // though not needed
            val paint = Paint()
            paint.color = backgroundColor
            paint.alpha = 255 //set the alpha
            paint.style = Paint.Style.FILL
            this.width = width + 20
            this.height = height
            fontSize = width / columnSize
            constructImage()
            initialise()
            //initalise the textposiotn to zero
            textPosition = IntArray(columnSize + 1) //add one more drop
            for (x in 0 until columnSize) {
                textPosition[x] = 1
            }
            drawFrame()
            changePicture()
            canvas!!.drawRect(0f, 0f, width.toFloat(), height.toFloat(), paint)
        }

        //old matrix effect code
        private fun drawText() {
            //loop and paint
            for (i in textPosition.indices) {
                // draw the text at the random position
                val y = textPosition[i]
                if (i >= 99 || y >= imageHeight * imageMultiply || foregroundImage[y][i] == 'ㅡ') {
                    textPaint!!.color = Color.parseColor("#00c800")
                    canvas!!.drawText(
                        "" + textChar[rand.nextInt(textLength)],
                        (i * fontSize).toFloat(), (y * fontSize).toFloat(),
                        textPaint!!
                    )
                    //
                } else {
                    textPaint!!.color = red
                    //canvas.drawText(".", x * fontSize, y * fontSize, paint);
                    canvas!!.drawText(
                        "" + foregroundImage[y][i],
                        (i * fontSize).toFloat(), (y * fontSize).toFloat(),
                        textPaint!!
                    )
                }

                // check if text has reached bottom or not
                //0.925 good
                if (textPosition[i] * fontSize > height && Math.random() > 0.900) //0.975
                    textPosition[i] =
                        0 // change text position to zero when 0 when text is at the bottom
                textPosition[i]++ //increment the position array
            }
        }

        private fun constructImage() {
            asciiString = getString(imgList[mImgIndex])
            imageHeight = (asciiString as String).length / 100
            foregroundImage = Array(imageHeight * imageMultiply) {
                CharArray(
                    100
                )
            }
            var modY = 0
            for (y in 0 until imageHeight) {
                for (x in 0..99) {
                    foregroundImage[modY][x] = (asciiString as String)[x + y * 100]
                    foregroundImage[modY + 1][x] = foregroundImage[modY][x]
                    //    foregroundImage[modY+2][x] =  foregroundImage[modY][x];
                }
                modY += imageMultiply
            }
        }

        //old martix effect code
        private fun canvasDraw() {
            //Log.d("canvas ","drawing");
            //set the paint for the canvas
            canvas!!.drawRect(
                0f,
                0f,
                width.toFloat(),
                height.toFloat(),
                backPaint!!
            ) //draw rect to clear the canvas
            drawText() // draw the canvas
        }
    }
}
